_posts/2021-04-22-torch-for-optimization/torch-for-optimization.Rmd

```diff
@@ -0,0 +1,640 @@
+---
+title: "torch for optimization"
+description: > 
+  Torch is not just for deep learning. Its L-BFGS optimizer, complete with Strong-Wolfe line search, is a powerful tool in unconstrained as well as constrained optimization.
+author:
+  - name: Sigrid Keydana
+    affiliation: RStudio
+    affiliation_url: https://www.rstudio.com/
+slug: keydanatorchoptimization
+date: 04-27-2021
+categories:
+  - Torch
+  - R
+output:
+  distill::distill_article:
+    self_contained: false
+    toc: true
+preview: images/preview.jpg
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+So far, all `torch` use cases we've seen have been in deep learning. However, its automatic differentiation feature is useful in other areas, as well. One prominent example is numerical optimization: We can use `torch` to find the minimum of a function.
+
+In fact, function minimization is *exactly* what happens in training a neural network. But there, the function in question normally is far too complex to even imagine finding its minima analytically. Numerical optimization aims at building up the tools to handle just this complexity. To that end, however, it starts from functions that are far less deeply composed. Instead, they are hand-crafted to pose specific challenges.
+
+This post is a first introduction to numerical optimization with `torch`. Central takeaways are the existence and usefulness of its L-BFGS optimizer, as well as the impact of running L-BFGS with line search. As a fun add-on, we show an example of constrained optimization, where a constraint is enforced via a quadratic penalty function.
+
+To warm up, we take a detour, minimizing a function "ourselves" using nothing but tensors. This will turn out to be less of a detour than expected though: Later, the overall process will still be the same. All changes will be related to integration of `optimizer`s and their capabilities.
+
+## Function minimization, DYI approach
+
+To see how we can minimize a function "by hand", let's try the iconic Rosenbrock function. This is a function of two variables:
```

Comment:
```suggestion
To see how we can minimize a function "by hand", let's try the iconic [Rosenbrock function](https://en.wikipedia.org/wiki/Rosenbrock_function). This is a function with two variables:
```

Reply:
Even if I'm the only one to read this post who _doesn't_ know what the Rosenbrock function is, it can't hurt to link out to it.

Reply:
yeah, I thought it would be fine without because I'm giving its definition as well as a picture, but you're right, a link can never hurt